### ISSUE TITLE
ML_1

### ISSUE THREAD

**Issue by DHRUVDALAL** [DHRUVDALAL/ml_codes#1]
# ============================================
# 📌 PCA on Wine Dataset to reduce dimensions
# ============================================

# Step 1: Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

# Step 2: Load the dataset directly from the URL
file_path = r"C:\\Users\\HP\\Desktop\\Wine.csv"   # <-- change this to your actual file path

df = pd.read_csv(file_path)

print("✅ Dataset Loaded Successfully")
print("Shape:", df.shape)
print("\n🔸 First 5 Rows:")
print(df.head())

# Step 3: Separate features (X) and target (y)
# 'Customer_Segment' is the target class (1, 2, 3)
X = df.drop('Customer_Segment', axis=1)
y = df['Customer_Segment']

print("\nFeature Columns:", list(X.columns))
print("Unique Target Labels:", y.unique())

# Step 4: Standardize the data (very important for PCA)
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Step 5: Apply PCA to reduce dimensions to 2 for visualization
pca = PCA(n_components=2)
X_pca = pca.fit_transform(X_scaled)

print("\n✅ PCA Transformation Complete")
print("Explained Variance Ratios:", pca.explained_variance_ratio_)
print("Total Variance Captured:", np.sum(pca.explained_variance_ratio_))

# Step 6: Create a new DataFrame with principal components
pca_df = pd.DataFrame(data=X_pca, columns=['PC1', 'PC2'])
pca_df['Customer_Segment'] = y

# Step 7: Visualize the PCA result
plt.figure(figsize=(10, 7))
for segment in pca_df['Customer_Segment'].unique():
    subset = pca_df[pca_df['Customer_Segment'] == segment]
    plt.scatter(subset['PC1'], subset['PC2'], label=f"Segment {segment}", s=50)

plt.xlabel('Principal Component 1')
plt.ylabel('Principal Component 2')
plt.title('PCA: Wine Dataset (2D Visualization)')
plt.legend()
plt.grid(True)
plt.show()

# Step 8: (Optional) Check how many components are needed to retain 95% variance
pca_full = PCA().fit(X_scaled)
cumulative_variance = np.cumsum(pca_full.explained_variance_ratio_)

plt.figure(figsize=(8,5))
plt.plot(range(1, len(cumulative_variance)+1), cumulative_variance, marker='o')
plt.xlabel('Number of Principal Components')
plt.ylabel('Cumulative Explained Variance')
plt.title('Explained Variance vs Number of Components')
plt.grid(True)
plt.axhline(y=0.95, color='r', linestyle='--', label='95% Variance Threshold')
plt.legend()
plt.show()
